### PR TITLE
add web_long_running_tests shard containing long-running web tests

### PR DESCRIPTION
**Pull request by yjbanov**
## Description

Add new shard `web_long_running_tests` containing long-running web tests. Initially contains 3 shards running the old gallery e2e test in the following modes:

- debug in HTML mode
- debug in CanvasKit mode
- profile in HTML mode
- profile in CanvasKit mode
- release in HTML mode
- release in CanvasKit mode

### CODE REVIEW

File: .cirrus.yml
Comment (nturgut) [inline .cirrus.yml:159]: why are we adding new tests to cirrus? I think we should add this one to LUCI, I added the details for this to go/flt-infra-web-long-tests

@godofredoc we are not adding new tasks to Cirrus CI now right, when is our ETA for flutter/flutter migration to LUCI?

Comment (nturgut) [reply .cirrus.yml:159]: Another reason I don't want the test in this location is the issue @yjbanov filed: https://github.com/flutter/flutter/issues/67569

As stated in Cirrus CI, Chrome is not hermetic. This tests depends on both chrome and chromedriver. It's best to have it LUCI, since on LUCI the builder is hermetic. We already have these dependencies ready.

File: dev/bots/test.dart
Comment (nturgut) [inline dev/bots/test.dart:798]: Web installers package does this.

In february we discussed that this functionality will be needed in 3 repos. That was the reason we first started web_installers repo. I think we should keep using it instead of duplicating code.

File: .cirrus.yml
Comment (godofredoc) [reply .cirrus.yml:159]: That is correct, migration of the SDK is complete. We are moving one last test this week and all the new tests should go directly to LUCI and no new tests should be added to Cirrus.

Comment (godofredoc) [reply .cirrus.yml:159]: This is also already running on LUCI https://cs.opensource.google/flutter/recipes/+/master:recipe_modules/adhoc_validation/resources/web_smoke_test.sh are reason to added to Cirrus?

File: dev/bots/test.dart
Comment (yjbanov) [reply dev/bots/test.dart:794]: Happy to switch to web_installers when it's ready. Currently there's no way to add a dependency on web_installers. We could use a pub dependency, but web_installers is not on pub. We could git clone it, but that's not how flutter/flutter manages its pub dependencies and it's a lot more code to make the git cloning work smoothly.

Comment (nturgut) [reply dev/bots/test.dart:794]: Why not, we are doing that since February? We are adding the dependency as SHA (as we are doing it with goldens).

https://github.com/flutter/engine/blob/master/lib/web_ui/pubspec.yaml#L35

(Actually going back to February, both creating web_installers and using it like this was your ides. You said you didn't want to create a copy code again (as we did for Chrome), so insisted that we should definitely have a web_installers repo for driver. I initially wanted to name that repo as "web_drivers" and you said Chrome should also be there. but never had a chance to carry Chrome :( )

File: .cirrus.yml
Comment (yjbanov) [reply .cirrus.yml]: I changed it. This is now a new shard called "web_long_running_tests" initially containing 3 subshards. PTAL.

File: dev/bots/test.dart
Comment (yjbanov) [reply dev/bots/test.dart:794]: In short, the flutter/engine method doesn't work in flutter/flutter due to the single-version rule policy and avoidance of git dependencies (the `flutter update-packages --force-upgrade` does not understand them leading to issues).

If web_installers were in pub, we could use it here. I'm not sure if it's worth it though. All I'm trying to do is launch `chromedriver`, which is already in PATH, unless it's already running. If web_installers improves this over the code as written in the PR, I'm all for switching. We should be mindful, however, that this will add a dependency, which has costs, such as having to send multiple PRs and going through the `flutter update-packages --force-upgrade` process in order to fix bugs.

Comment (nturgut) [reply dev/bots/test.dart:794]: As you mentioned in the offline discussion, if you believe "circumstances changed therefore we don't need a common repository" feel free to use this code. I personally favor common code usage. Especially considering this long running test will need:
- firefox driver code
- ios-safari simulator code
-safari driver code
in its lifetime, I feel like this class will be a very big copy/paste location. 

Note that if change something in the future (add verbose to drivers,  run on a different port), this is the only driver starting code outside of the `web_installers` common package. Therefore forgetting this will be easier. 

Also @godofredoc has a plan for separating tests/builds/dependency steps. When we do that,  this type of code (for example related to chrome driver or ios-simulator) will be in another step. When we look at a LUCI result, we will be able to clearly see "is tests failed", "is build failed", "is setup failed". 

I was hoping to use web_installers directly in that pre-steps. In this case looks like we would need to modify this class in the future.

If you want to use this code you can document these discussions to comments so if one webdriver tests fails and all the others are passing the developers who do the firefighting can clearly see how this one is different.

File: .cirrus.yml
Comment (nturgut) [reply .cirrus.yml]: great! I added more comments to the how to doc, let me know if you need mode details.

**Review by nturgut — COMMENTED** [review]
Looks like some builds are failing. I assume we still need to add more to the PR. I'll have another look once the build steps are working.

File: dev/bots/test.dart
Comment (nturgut) [reply dev/bots/test.dart:793]: Ah, I'm sorry, while checking the code I realize this part of the comment is not true: "Note that if change something in the future (add verbose to drivers, run on a different port), this is the only driver starting code outside of......". There are other places which use a stand alone chromedriver command.

Comment (yjbanov) [reply dev/bots/test.dart:793]: The risk of code duplication makes sense. Would you be able to publish `web_installers` to pub so we can reuse it across all repos like a regular pub dependency?

Comment (nturgut) [reply dev/bots/test.dart:793]: Can you create an issue? I'm working on other P3 but I will try prioritize it.

Comment (yjbanov) [reply dev/bots/test.dart:793]: I will. In the meantime, can you PTAL at the PR again?

**Review by nturgut — COMMENTED** [review]
Does this test uses MouseEvents? If so you might want to wait for this fix to get it: https://github.com/flutter/plugins/pull/3168

File: dev/bots/run_command.dart
Comment (nturgut) [inline dev/bots/run_command.dart:214]: very nit: the command output?

File: dev/bots/run_command.dart
Comment (nturgut) [inline dev/bots/run_command.dart:229]: info for later use: If we want to add screenshots to these tests (and still want to use flutter_goldens, for example for flutter/engine repo tests) we read the log_paths for outputting the failed golden information. Example from web engine tests: https://flutter.googlesource.com/recipes/+/refs/heads/master/recipes/web_engine.py#201

Does the capture mode swallow the output, or will they still be in the logs?

File: dev/bots/test.dart
Comment (nturgut) [reply dev/bots/test.dart:793]: Sure, I reviewed it all now :) Thanks for adding me as a reviewer.

File: dev/bots/test.dart
Comment (yjbanov) [reply dev/bots/test.dart:79]: Shard count is not passed to the test script, only the shard index. E.g. when you run it as `SHARD=web_tests SUBSHARD=4 dart dev/bots/test.dart` the script needs to know if it's 4 out of 10, or 4 out of 20. So total shard counts are hard-coded in the test script. See also `kBuildTestShardCount`, `webShardCount`, and usages of `selectSubshard`.

File: dev/bots/test.dart
Comment (yjbanov) [reply dev/bots/test.dart:755]: It will be one shard, but it will be split into `kWebLongRunningTestShardCount` subshards, which we can tweak over time. The setup we use for the `web_tests` shard is ideal for this use-case. `web_integration_tests` doesn't have subshards, so it won't scale.

File: dev/bots/test.dart
Comment (yjbanov) [reply dev/bots/test.dart:825]: Done.

File: dev/bots/test.dart
Comment (yjbanov) [reply dev/bots/test.dart:840]: I don't have a strong preference. Specifying both makes it more explicit, I think.

File: dev/bots/run_command.dart
Comment (yjbanov) [reply dev/bots/run_command.dart:214]: Done, in https://github.com/flutter/flutter/pull/68324 (this PR contains changes from the other PR so I could test them)

File: dev/bots/run_command.dart
Comment (yjbanov) [reply dev/bots/run_command.dart:229]: `capture` means store the output in memory and return to the caller. It's up to the caller to decide what to do with it. They could drop it on the floor, print to the logs, inspect it, or any combination of these.

File: dev/bots/test.dart
Comment (nturgut) [reply dev/bots/test.dart:755]: web_tests are setup to shards from config file as well. I might be missing something but I didn't get the usage for `kWebLongRunningTestShardCount`. Also the comment might also be misleading since the number we set to the config will determine the number of subshards not the constant we add here.

```
/// These tests are sharded into [kWebLongRunningTestShardCount] shards.
```

Note for later: I added another recipe to the engine side for running web_tests. we might still need to make some modifications on that.

**Review by nturgut — APPROVED** [review]
Have you tested this PR with `led`? We can also use `led` on unmerged PR so it's better to test this before merge this one.

You can invoke this builder https://github.com/flutter/infra/blob/e6d2883f0fdd81d653c3ec0ed554a477a09bb9ec/config/framework_config.star#L181 with correct shard/subshard properties.

I added a lot examples here: go/fw-recipe-manual for using a builder with a PR and changing properties. If tests passes when running with `led` I think we can merge the PR.

File: dev/bots/test.dart
Comment (yjbanov) [reply dev/bots/test.dart:755]: Check out the line below: `await _selectIndexedSubshard(tests, kWebLongRunningTestShardCount);`. `_selectIndexedSubshard` uses the count to divide the total number of tests (6) into subshards (3) then run one of them (2 tests per subshard). The hard-coded count is needed because `test.dart` doesn't receive it from LUCI configs (i.e. the number 3). Started a discussion on #hackers-infra.

Comment (nturgut) [reply dev/bots/test.dart:755]: I see thanks for the explanation. For using the same value with LUCI infra we have 2 options:

-> We can use the value from flutter/infra as source of truth. We modify the recipes to write an env variable. This code reads that env variable.
-> We can write it to a yaml file in this repo (so the source of truth is flutter/flutter). We can read the yaml file from the LUCI recipe (we need to be careful with this design since flutter.py and flutter_drone.py is very generic, we don't want to add a random if/else there).

Both options can easily be tested with `led`.

@godofredoc @christopherfujino which do you think would be the best approach.

Comment (godofredoc) [reply dev/bots/test.dart:755]: I'd vote for adding the value to flutter/infra and use it in the script. As part of the LUCI transition we would like to keep everything related to infra resources allocation/usage in the LUCI space to have more visibility and better management of the resources.

File: dev/bots/test.dart
Comment (nturgut) [reply dev/bots/test.dart:81]: It may be because, the shard count is not passed that way to builders.

I usually refer to already running builders to understand how properties work, for example: https://ci.chromium.org/p/flutter/builders/prod/Linux%20web_tests/878

Shards is:  "web_tests", subshards is: [ "0", "1", "2", "3", "4", "5", "6", "7_last" ]

When these are provided is there an information not set by the recipe?

Recipe code is here btw: https://flutter.googlesource.com/recipes/+/refs/heads/master/recipes/flutter/flutter_drone.py

**Comment by yjbanov** [comment]
Here's a successful run from `led` for each subshard:

- [subshard 0](https://ci.chromium.org/p/flutter/builders/try/Linux%20SDK%20Drone/90127?)
- [subshard 1](https://ci.chromium.org/p/flutter/builders/try/Linux%20SDK%20Drone/90128?)
- [subshard 2_last](https://ci.chromium.org/p/flutter/builders/try/Linux%20SDK%20Drone/90135?)

This run includes @jonahwilliams' `flutter drive` fix, so these should not be false positives.

File: dev/bots/test.dart
Comment (yjbanov) [reply dev/bots/test.dart:81]: > When these are provided is there an information not set by the recipe?

Yes, the recipe only sets `SHARD` (e.g. `web_tests`) and `SUBSHARD` (e.g. `3`) environment variables: https://flutter.googlesource.com/recipes/+/refs/heads/master/recipes/flutter/flutter_drone.py#55. The recipe does not pass the total number of available shards. That's the missing information.

File: dev/bots/test.dart
Comment (yjbanov) [reply dev/bots/test.dart:789]: This sounds like something outside the scope of my PR. However, I'd vote for adding these to flutter/flutter. Currently flutter/infra leads to non-hermetic builds (you build a PR with 5 subshards and see subshard 3 fail; you build it again but someone changed the shard count to 8 in flutter/infra, and now subshards 1 and 6 fail; you pull your hair).

If the system is designed correctly, adding new shards or subshards should not affect infra resource allocation. Instead a PR should get up to a limited amount of resources (the limit can change depending on the overall system load), and schedule the shards/subshards (no matter how many of them) on top of the available resources, similar to how an OS will schedule processes/threads on a limited number of CPU cores, no matter how many processes/threads there are, and similar to how Bazel schedules thousands of build targets with concurrency limited to your CPU core count.

However, if the infra is unable to implement per-PR scheduling yet, we could require that the file that controls sharding is reviewed by someone from the infra team.

**Comment by nturgut** [comment]
> Here's a successful run from `led` for each subshard:
> 
> * [subshard 0](https://ci.chromium.org/p/flutter/builders/try/Linux%20SDK%20Drone/90127?)
> * [subshard 1](https://ci.chromium.org/p/flutter/builders/try/Linux%20SDK%20Drone/90128?)
> * [subshard 2_last](https://ci.chromium.org/p/flutter/builders/try/Linux%20SDK%20Drone/90135?)
> 
> This run includes @jonahwilliams' `flutter drive` fix, so these should not be false positives.

Nice! we can merge this one after the tests are passing. Then we need to merge the infra PR, the update the try/prod json files.

**Pull request merged**
